### DISCUSSION TITLE
[JS] What to do with `[[identifiers]]`, _internal properties and methods_ of JS objects?

### DISCUSSION THREAD

**Discussion by teoli2003** [mdn/mdn-community#150] [Content]
### The problem

JavaScript objects have what are called _internal properties and accessors_. These are values and functions stored on a JavaScript object, but, unlike regular properties and functions, these entities cannot be read, set, or called by any code outside _the code of the engine_ itself.

In the specification, they are present all over the place, and on MDN, we find quite a few mentions of them. We recognize them as they are enclosed between two sets of brackets, like `[[prototype]]`, `[[value]]`, or `[[set]]`.

The problem is the following: these mentions are information that comes out of the spec, this is not something that web developers directly deal with. At first sight, it looks like something we should bury and not display to web devs. On the other side, if authors have used them here or there (and this is not recent!), it means they are solving a problem: maybe they are more precise and succinct, and maybe they are needed to describe complex interactions.

The current situation is not good: we use these notations here or there without really explaining them, making it hard for a lot readers. We should find a solution that allows readers to naturally understand the docs (either by explaining better the internal properties when used or by hiding them). 

Right now, I don't what is the best direction to take.

### What questions do we want to answer in this discussion?

With this discussion, I have three objectives:

1. Understand _why_ we use them sometimes on MDN. 
2. _What_ do we want to do with them in the future (more of them, them differently, or never?).
3. A brief plan about _how_ to reach the ideal situation.

To help us make pertinent proposals, here is some extra information.

### What does the spec say?

There are two kinds of them: _internal methods_ and _internal slots_.

> The actual semantics of objects, in ECMAScript, are specified via algorithms called **internal methods**. Each object in an ECMAScript engine is associated with a set of internal methods that defines its runtime behaviour. These internal methods are not part of the ECMAScript language. They are defined by this specification purely for expository purposes. However, each object within an implementation of ECMAScript must behave as specified by the internal methods associated with it. The exact manner in which this is accomplished is determined by the implementation.
>
> **Internal method** names are polymorphic. This means that different object values may perform different algorithms when a common internal method name is invoked upon them. That actual object upon which an internal method is invoked is the “target” of the invocation. If, at runtime, the implementation of an algorithm attempts to use an internal method of an object that the object does not support, a TypeError exception is thrown.
>
> **Internal slots** correspond to internal state that is associated with objects and used by various ECMAScript specification algorithms. Internal slots are not object properties and they are not inherited. 

([spec](https://tc39.es/ecma262/#sec-object-internal-methods-and-internal-slots))

### How often are they used on MDN today?

| Internal prop or accessor |  Usage (total/pages)    | Significant links                              | Meaning                                                 | In Spec |
|--------------------------|:--------:|----------------------------------|--------------------------------------|:-------:|
| [[Prototype]]                     | 68 / 17 |                                                         | Internal slot containing the _prototype_ of an object, or `null`. |  [Yes](https://tc39.es/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots)             |
| [[Get]]                               | 4 / 4     | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/object/assign) [2](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/proxy/proxy/get) [3](https://developer.mozilla.org/en-us/docs/web/javascript/data_structures) [4](https://developer.mozilla.org/en-us/docs/web/javascript/guide/meta_programming)                       |  The _getter_ associated with the object.                                    |[Yes](https://tc39.es/ecma262/#table-essential-internal-methods)|
| [[Set]]                                | 4 / 4    | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/object/assign) [2](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/proxy/proxy/get) [3](https://developer.mozilla.org/en-us/docs/web/javascript/data_structures) [4](https://developer.mozilla.org/en-us/docs/web/javascript/guide/meta_programming)                       | The _setter_ associated with the object.                                                               |[Yes](https://tc39.es/ecma262/#table-essential-internal-methods)|
| [[foo]]                                | 2 / 2     | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/proxy/revocable) [2](https://developer.mozilla.org/en-us/docs/web/javascript/guide/meta_programming)                                        |The two occurrences are in comments in a code snippet that uses the double bracket notation seemingly to list propeties (and so not really à propos)|No|
| [[Construct]]                    |  2 / 2     | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/proxy/proxy/construct) [2](https://developer.mozilla.org/en-us/docs/web/javascript/reference/operators/spread_syntax)                                        | Internal method mandatory for _constructors_. Creates an object. Invoked via the `new` operator or a `super()` call. | [Yes](https://tc39.es/ecma262/#table-additional-essential-internal-methods-of-function-objects)|
| [[Call]]                              | 4 / 3     | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/function/bind/) [2](https://developer.mozilla.org/en-us/docs/web/javascript/reference/operators/typeof) [3](https://developer.mozilla.org/en-us/docs/web/javascript/reference/operators/spread_syntax)                                | Internal method mandatory for _functions_ (and _constructors_). Executes code associated with this object. Invoked via a function call expression. | [Yes](https://tc39.es/ecma262/#table-additional-essential-internal-methods-of-function-objects)|
| [[TimeValue]]                   | 1 / 1             | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/date/tostring)                                                 |Explain that `toString()`is an abstract method that can only be used on `Date` objects as they need a `[[TimeValue]]` internal slot. This doesn't seems to be true in ES2018 and the current draft.  |  No         |
| [[Put]]                               | 2 / 2      | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/typedarray/of) [2](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/typedarray/from)                                        |Two occurrences to control as this is unclear and not in the spec (anymore?)|No              |
| [[DefineProperty]]           | 2 / 2      | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/typedarray/of) [2](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/typedarray/from)                                        |Called `[[DefineOwnProperty]]`. Internal function for all objects called to create or alter an own property of the object.                                                       |[Yes](https://tc39.es/ecma262/#table-essential-internal-methods)|
| [[Define]]                         | 2 / 1      | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/classes/public_class_fields)                                                 | Likely an erroneous abbreviation of `[[DefineOwnProperty]]`.|No|
| [[OwnPropertyKeys]]      | 1 / 1               | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/proxy/proxy/ownkeys)                                                 |Internal slot for all objects with the list of _own properties_.                           |[Yes](https://tc39.es/ecma262/#table-essential-internal-methods)|
| [[Key]]                               | 1 / 1               | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/symbol/for)                                                |Global internal structure: the _global symbol registry_, a map of symbols associated with a key. | [Yes](https://tc39.es/ecma262/#sec-symbol.for) |
| [[symbol]]                        | 1 / 1               | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/symbol/for)                                                 |Internal slot of the _global symbol registry_, a map of symbols associated with a key. |[Yes](https://tc39.es/ecma262/#sec-symbol.for)|
| [[PromiseStatus]]            | 2 / 1      | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/promise/then)                                                 |Called `[[PromiseState]]`. An internal slot containing the _status_ of the promise, a state machine: `pending`, `fullfilled`, or `rejected`. **Note:** This state machine is [described](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#description) on MDN, but doesn't use the notation `[[PromiseState]]`…|[Yes](https://tc39.es/ecma262/#table-internal-slots-of-promise-instances) |
| [[PromiseValue]]              | 2 / 1      | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/promise/then)                                                 |Called [[PromiseResult]]. An internal slot with the value with which the promise has been fulfilled or rejected, if any. **Note:** Both `[[PromiseResult]]` and `[[PromiseState]]` are used on MDN only in a code comment about what should be displayed on the console. This comment is outdated and doesn't describe what Chrome or Firefox dev tools does (and they display different texts).|[Yes](https://tc39.es/ecma262/#table-internal-slots-of-promise-instances)|
| [[VarNames]]                   | 2 / 1      | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/statements/var)                                                 |Global internal structure. A list of string names bound by functions, generators, and var declaration in global code |[Yes](https://tc39.es/ecma262/#table-additional-fields-of-global-environment-records)|
| [[BoundThis]], [[BoundTargetFunction]], [[BoundArguments]]                 | 2 / 1      | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/function/bind/)                                                 |The 3 internal slots of a _bound function exotic object_. **Note:** On MDN, these 3 values are used to paraphrase what implementations do behind the scene.|[Yes](https://tc39.es/ecma262/#table-internal-slots-of-bound-function-exotic-objects)|
| [[Multiline]]                      | 1 / 1               | [1](https://developer.mozilla.org/en-us/docs/web/javascript/reference/global_objects/regexp/sticky)                                                | An internal slot of `RegExp`, a boolean storing if the multiline (`/m`) flag is set or not. **Note:** On MDN, this appears as part of a citation of the spec.|[Yes](https://tc39.es/ecma262/#table-regexp-record-fields)|

cc/ @hamishwillee @rumyra @wbamberg @sideshowbarker @josh-cena @schalkneethling

**Comment by teoli2003** [comment]
While researching to understand and describe the problem, it looked to me that the usage of this notation is uncommon, complex to understand, and not needed.

### TL;DR;

**We should not use the `[[identifier]]` notation on MDN at all**.

(With the exception of pages with the specification as their topic, like a page about _How to document a new JavaScript feature_).

### Reasoning

- Most of the time, an entity described by the notation is used at a single place on MDN. Introducing a notation for a single usage complexifies the documentation.
- The notation is not understood by most developers.
- Like for Web APIs' mixins, there is no guarantee of stability: naming can change, and there is no process in place to update this.
- The identifiers are invisible to developers.
- There is no explanation of such notation on the web, outside the EcmaScript spec (and even there, there isn't a true explanation)
- The sections using them are not very clear and would benefit from a rewrite to be easier to understand.

### Plan

1. Update the pages using them so we get rid of them. Eventually uniformizing some wordings.
2. Write a page about _How to write documentation for a new JS feature_ to guide authors not to use them.

**Comment by Josh-Cena** [comment]
IMO—they are quite unavoidable in certain places.

For example, when you say "the method is defined on constructor's prototype", does that mean `constructor.__proto__`, or `constructor.prototype`? In the case you are referring to the former, it's really bad notation because `__proto__` accessors are deprecated; but otherwise there's no way to refer to `[[Prototype]]`. `Object.getPrototypeOf(constructor)` is definitely too verbose to be illustrative. That's why [[Prototype]] is used extensively on the https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain page, because there's really no other non-ambiguous notation.

Also, `[[Define]]` (yes, short for [[DefineOwnProperty]] but it's quite common AFAICT) and `[[Set]]` are not directly accessed, but their semantics are exposed by assignments and `Object.defineProperty`, and they can be hooked into by setters and property descriptors. For example, when we document class fields, the knowledge that it uses `[[Define]]` may be gibberish to average readers, but to those that do understand, it has much, much further implications than "it doesn't invoke setters" (e.g. its interaction with writability and configurability), and this single piece of information is worth 5 paragraphs.

(The consequence of avoiding `[[Define]]` is some dangerously misleading text, like "They are added to the class constructor at the time of class evaluation using [Object.defineProperty()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty)." How should I understand that? Does that mean if I overwrite `Object.defineProperty` I overwrite how class fields are defined as well?)

This is more personal, but I still want to share it: the first time I started reading MDN systematically (i.e. not as a lookup tool for random APIs) was when I was implementing a [klass](https://github.com/jc-verse/ts-klass)—an API that tries to mimic every single behavior of JS classes. As such I needed to know every single aspect of the language's behavior, but which was surprisingly lacking on MDN. Of course 99% of our readers would not need to know what happens when you use class fields with a property already defined as non-configurable, but for those that do want to know, do we want to explain it in such a roundabout way than simply stating "it uses [[Define]]", which saves me 1 precious minute reading through the explanation and code examples?

Can we instead create a page under JS references > misc like "Exotic objects and internal slots", where we document these internal slots, and then everywhere we use them, we link to this page? There are other interesting behaviors with internal slots, e.g. you cannot do `new Proxy(new Map)`, because the proxy cannot get access to the map's `[[MapData]]` internal slot.

**Reply by hamishwillee** [reply]
Who is our audience? Is it the 99% or the 1%? Or both?

The proposal to have a "Exotic objects and internal slots" doc as a linkable reference sounds pragmatic. But I'd very much prefer it was treated as a truly last resort. Even it it costs us as maintainers more work.

**Reply by Josh-Cena** [reply]
Both.

MDN has an incredibly diverse audience and I would try as hard as possible to please both ends. We have TC39 folks, library authors, and compiler implementors all reading our docs to gain some insight of the language's behavior, because the spec _describes_ them but doesn't really _summarize_ them. We really lack high-end systematic materials in the ecosystem—there's an unbridgeable gap between documentation and the spec. Using precise terms like "uses `[[DefineOwnProperty]]`" instead of some awkward "uses `Object.defineProperty`" (or not pointing that out at all but re-iterating all implications of `[[DefineOwnProperty]]` with prose), paired with a linked explanatory page, is IMO a net gain.

**Reply by Josh-Cena** [reply]
(Speaking of internal slots + proxies, I've created https://github.com/mdn/content/pull/18209 to illustrate how internal slots are actually observable in JS code.)

**Reply by hamishwillee** [reply]
@Josh-Cena That's a very good explanation ^^^. I'm happy to target both - like I said, your proposal sound pragmatic. But let's structure for the majority.

**Reply by jpmedley** [reply]
I'm answering as someone who's spent 7 years working with Chrome developers to document features on MDN that they've added to Chrome . Understand that their job is to describe features in under-the-hood terms so that fellow engineers know what to implement. I'd argue that much of this stuff has likely crept into MDN because it didn't occur to contributors that features should be described any other way. Before the move to gitHub we had no way to stop this. I don't know about the rest of you, but I always ask the for removal of internals from content that I review. If I were to review something that's hard to understand without reference to internals, I'd go file an issue in the spec's repo and I'd encourage anyone else to do the same.

Contributors also tend to copy what they find on a single page without asking whether something is supposed to be on the site at all. In other words, I suspect the majority of existing browser-internal explanations are the result of default behavior, not deliberate choices.

Question: doesn't good API design require a clear set of inputs, a clear set of outputs that proceed logically from the inputs, and NO side effects? That question brings me to two conclusions. First, if a platform feature is well-designed, it should be describable without reference to internals. Some features may not meet this standard, so, second, reference to internals should be the exception, not the rule. In any case, we always link to specs for anyone who's curious about the internals and needs to see them for some reason.

**Reply by Josh-Cena** [reply]
> doesn't good API design require a clear set of inputs, a clear set of outputs that proceed logically from the inputs, and NO side effects? 

Absolutely—however as I've demonstrated above the _existence_ of internal slots is not entirely invisible to the user. Also, the ones I'm talking about ([[Define]], [[Prototype]], etc.) are not really internal state (as I guess what you mean by side-effects)—they are more like "behavior" that can be captured in JS in various ways. But because these behaviors don't always have their trivial JS equivalent, reference to the internal terminology seems the least ambiguous. (For example, if `__proto__` was ever standardized I'd be more than willing to replace [[Prototype]] with `__proto__`)

Again, I'm all in for removing _excessive_ reference to internal details where such loss of abstraction is pointless. For example, on this page: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind, the references to `[[BoundTargetFunction]]` et al. are basically pointless because they can be equally described with prose (something like "the bound `this` and arguments are stored internally in the bound function similar to a closure"). So yes, I agree with the stance that "reference to internals should be the exception, not the rule.". But then there are other places where it's IMO unavoidable. I'm currently holding on to [[Define]], [[Set]], and [[Prototype]] as the exceptions that I would tolerate (maybe [[Call]] as well, considering the `typeof` page needs to mention it to explain what things would return `"function"`).

**Reply by hamishwillee** [reply]
It would be much easier to talk about this if we eliminated all the cases that we're in agreement on. Then we're left with a much smaller set of cases where we can test the theory that it makes sense to keep them.

**Reply by Josh-Cena** [reply]
I agree. I've started with https://github.com/mdn/content/pull/18461 — we can gradually remove these.

**Reply by teoli2003** [reply]
I agree; let's start by removing the case where we are in agreement. For the other cases (I think there are two of them), we can look at them later.

**Reply by Rumyra** [reply]
> I agree; let's start by removing the case where we are in agreement. For the other cases (I think there are two of them), we can look at them later.

This sounds like a good way to move forward :)

It looks like the ones up for discussion are:
* [[Define]]
* [[Set]]
* [[Prototype]]
* Possibly [[Call]]

**Reply by teoli2003** [reply]
Not sure for [[Set]], but we may have [[Put]] instead.

**Reply by Josh-Cena** [reply]
`[[Put]]` is ES5—it predates `[[Set]]`. So yes, [[Define]] (or [[DefineOwnProperty]] to be exact), [[Set]], and [[Prototype]] are my "yes"; [[Call]] and [[Construct]] are my "maybe".

**Comment by Josh-Cena** [comment]
After https://github.com/mdn/content/pull/18461 and https://github.com/mdn/content/pull/18476, we have gotten rid of some pointless references. @teoli2003 Maybe you can update that table? I think the rest are more debatable. (The proxy ones are tricky—they are not very critical, but I suspect people who read these pages have a fair chance of understanding the notation anyway.)
